### ISSUE TITLE
Advanced Standing and Preparation For: Enable third-party assertions/mapping

### ISSUE THREAD

**Issue by jkitchensSIUC** [CredentialEngine/vocabularies#392]
Per State of Indiana.

- Must be able to show credit hours/units of any type that can be transferred for advanced standing.
- Must be able to show a description.
- Must be able to show exceptions to the credit hours/units of any type that can be transferred for advanced standing.
- Must be able to refer to a program or designation that guarantees a credit transfer.
- Would be helpful to show how this is as a time reduction equivalent.
- Must be able to show military transfer application.
- Identify the connection between a military occupation code (https://www.onetonline.org/crosswalk/MOC/)  and the credential(s) for which it grants preparation or advanced standing

The American College on Education is working with Military and State's postsecondary approval bodies on recommendations for military occupations and training as advanced standing and preparation for civilian credentials.  

The CTDL Editor will have to be updated ahead of the June release to support this for this statewide project of the Credential Engine.

Indiana Google Folder https://drive.google.com/drive/folders/0B1uSTaYpjhXJUTVSa1V6eDYtdDQ?
usp=sharing
See this Google Sheet https://docs.google.com/spreadsheets/d/1cbyxfTCVq3BiskChnEfG-SBEZVis9wfqIcdQzQLFt2s/edit?usp=sharing

**Comment by cwd-mparsons** [comment]
@jkitchensSIUC 
I added a section on this request to an editor work plan document.
https://docs.google.com/document/d/1JFzhmKmZbxA4H-tQTC29qfwqUxHwIujjuhC7h5Un-M8/edit#heading=h.9fn0h64pgq9e
Following are some of the questions I have at this time. These were copied from the latter document.
**Credit Hour Exceptions**
- How to show exceptions?

**Refer to a program or designation that guarantees a credit transfer**
- Is this a text property?
- Or a learning opportunity
What are examples of designations
- Will these be literals, or a vocabulary

**Show how this is as a time reduction equivalent**
What is a time reduction equivalent

**Must be able to show military transfer application**
What are the steps or properties necessary to show military  transfer application?

**Comment by stuartasutton** [comment]
Jeanne, before there are editor updates, we need to walk through these requirements to determine whether new properties are needed to the CTDL.

**Comment by siuc-nate** [comment]
Notes from our 5/2/2017 call:
- Are details about connections between credentials resolved through condition profile or do they require their own profile?
- How do/can we enable third parties to publish assertions about the relationships between two credentials when such parties are not the owner of either credential?
   - Example: State assertions about a credential providing advanced standing for another credential

**Comment by stuartasutton** [comment]
@siuc-nate, it would seem to me that for your bullet one that condition profile should work for the kinds of assertions made by the creator of resource that has conditional relationships with a second resource created by someone else. It might need a new property or two to handle the Indiana use cases (would be really nice to see some real Indiana examples so this discussion isn't so abstract).

Your second bullet covering assertions by a 3rd party of an association between two resources created by one or two other parties is a totally different matter and has a CTDL aspect and a separate CER aspect. The CER has policy (should we/shouldn't we support) and technical constraints.  

The CTDL aspect --to the extent we want to support the use case-- would likely involve a new 3rd-party "mapping" entity not unlike the IMS CASE "Association" which would come with certain dangers in practice (main problem would be a **really bad practice** of using it as a substitute for all direct properties turning a graph model into a relational model). I'll prep a diagram and a couple of coding examples of 3rd party use of such a beast.

**Comment by siuc-nate** [comment]
Some ideas to handle third party assertions, based on our call today:

Very basic class:
```
{
   "@id": "http://www.credentialengineregistry.org/resources/ce-[UUID]",
   "@type": "ceterms:ThirdPartyAssertion",
   "ceterms:assertedBy": { "@id": "http://www.credentialengineregistry.org/resources/ce-[UUID]" },
   "ceterms:assertionSubject": { "@id": "http://www.credentialengineregistry.org/resources/ce-[UUID]" },
   "ceterms:assertionPredicate": { "@id": "http://www.credentialengineregistry.org/resources/ce-[UUID]" },
   "ceterms:assertionObject": { "@id": "http://purl.org/ctdl/terms/isAdvancedStandingFor" }
}
```
As you might guess, "assertionSubject/Predicate/Object" are equivalent to rdf:subject/predicate/object, respectively.  

It might not hurt to enable a little more detail:
```
{
   "@id": "http://www.credentialengineregistry.org/resources/ce-[UUID]",
   "@type": "ceterms:ThirdPartyAssertion",
   "ceterms:name": "My advanced standing assertion",
   "ceterms:description": "The [credential 1] credential grants advanced standing for [credential 2].",
   "ceterms:assertedBy": { "@id": "http://www.credentialengineregistry.org/resources/ce-[UUID]" },
   "ceterms:assertionSubject": { "@id": "http://www.credentialengineregistry.org/resources/ce-[UUID]" },
   "ceterms:assertionPredicate": { "@id": "http://www.credentialengineregistry.org/resources/ce-[UUID]" },
   "ceterms:assertionObject": { "@id": "http://purl.org/ctdl/terms/isAdvancedStandingFor" }
}
```
But then that also opens the door to questions of "what about weight?" or "what about credit hours?" and so on, which can very quickly balloon to the equivalent of "what about ConditionProfile?".  Rather than rebuild ConditionProfile one property at a time, we could just include it:
```
{
   "@id": "http://www.credentialengineregistry.org/resources/ce-[UUID]",
   "@type": "ceterms:ThirdPartyAssertion",
   "ceterms:name": "My advanced standing assertion",
   "ceterms:description": "The [credential 1] credential grants advanced standing for [credential 2].",
   "ceterms:assertedBy": { "@id": "http://www.credentialengineregistry.org/resources/ce-[UUID]" },
   "ceterms:assertionSubject": { "@id": "http://www.credentialengineregistry.org/resources/ce-[UUID]" },
   "ceterms:assertionPredicate": { "@id": "http://www.credentialengineregistry.org/resources/ce-[UUID]" },
   "ceterms:assertionObject": { "@id": "http://purl.org/ctdl/terms/isAdvancedStandingFor" },
   "ceterms:assertionConditions": [
      {
         "@type": "ceterms:ConditionProfile",
         "ceterms:creditHourType": "Semester Hours",
         "ceterms:creditHourValue": 10,
         "ceterms:submissionOf": [ "Transcript", "Proof of residency" ],
         "ceterms:residentOf": [ 
         { 
            "@type": "ceterms:JurisdictionProfile",
            "ceterms:mainJurisdiction": [
               {
                  "@type": "ceterms:GeoCoordinates",
                  "ceterms:geoURI": "http://www.geonames.org/4921868"
                  "ceterms:name": "Indiana"
               }
            ]
         }
         ],
         "ceterms:weight": 9.0,
         "ceterms:audienceType": [
            {
               "@type": "ceterms:CredentialAlignmentObject",
               "ceterms:targetNode": "audience:Citizen" 
            },
            {
               "@type": "ceterms:CredentialAlignmentObject",
               "ceterms:targetNode": "audience:InState"
            }
         ],
         "ceterms:subjectWebpage": "http://indiana.us.gov/assertions/about/credentials"
      }
   ]
}
```
But then that leaves open questions like "Doesn't ConditionProfile also have a name and description and an assertedBy?" and "What about the target_____ properties?"

You're basically trying to provide the subject and connection type, and use ConditionProfile to carry out the rest.  So, you could do:
```
{
   "@id": "http://www.credentialengineregistry.org/resources/ce-[UUID]",
   "@type": "ceterms:ThirdPartyAssertion",
   "ceterms:assertionSubject": { "@id": "http://www.credentialengineregistry.org/resources/ce-[UUID]" },
   "ceterms:isAdvancedStandingFor": [
      {
         "@type": "ceterms:ConditionProfile",
         "ceterms:targetCredential": { "@id": "http://www.credentialengineregistry.org/resources/ce-[UUID]" },
         "ceterms:assertedBy": { "@id": "http://www.credentialengineregistry.org/resources/ce-[UUID]" }
      }
   ]
}
```
But then you've literally just reinvented ConditionManifest and added the assertionSubject property. Considering that one of the benefits of ConditionManifest is that things not owned by its creator organization can reference it, it's not far off to think that ConditionManifest could be used to reference things its creator doesn't own, too.  Aside from needing some property to indicate the subject of the assertion, it seems to cover all of the other bases already.  

If just we create a "conditionSubject" property, add it to ConditionManifest (or maybe just reuse the target____ series of properties?), then I _think_ that would solve this problem.  

Or you could make ConditionProfile a top-level entity, add conditionSubject and conditionType properties, and end up with something like:
```
{
   "@id": "http://www.credentialengineregistry.org/resources/ce-[UUID]",
   "@type": "ceterms:ConditionProfile",
   "ceterms:conditionType": { "@id":  "ceterms:isAdvancedStandingFor" },
   "ceterms:conditionSubject": { "@id": "http://www.credentialengineregistry.org/resources/ce-[UUID]" },
   "ceterms:targetCredential": { "@id": "http://www.credentialengineregistry.org/resources/ce-[UUID]" },
   "ceterms:assertedBy": { "@id": "http://www.credentialengineregistry.org/resources/ce-[UUID]" },
   ...Other properties
}
```
Which feels cleaner at first, but then that allows something like this:
```
{
   "@type": "ceterms:Certificate",
   "ceterms:requires": [
      {
         "@type": "ceterms:ConditionProfile",
         "ceterms:targetCredential": { "@id": "http://www.credentialengineregistry.org/resources/ce-[UUID]" },
         "ceterms:conditionType": { "@id":  "ceterms:isAdvancedStandingFor" },
      }
   ]
}
```
Basically a mismatch between the property that points to ConditionProfile and the property within the ConditionProfile.

So I would lean towards recycling ConditionManifest ( _perhaps_ creating a ThirdPartyAssertion subclass of it instead?) and adding an assertionSubject property to it.

**Comment by stuartasutton** [comment]
@siuc-nate, this should be split out of this github issue. This entails a lot more than `ceterms:advancedStanding` -- the subject of this particular issue. It's actually a general 3rd party mapping entity. This is particularly true when we look at something like this in terms of relating two competencies, or as seen in the CASE demo yesterday, relating a learning resource to a competency. So, it's a much more expansive issue than just `ceterms:advancedStanding` and deserves a separate github issue.

First, let's not spiral off into the ozone with what might be. While it might be quite practical to subclass ConditionProfile and allow a map to express some conditions, we should hold even on that. We should let articulated need drive expansion. So, I'd start with your second example replicated here with two alterations: (1) switch incorrect values for predicate and object; (2) and actually use the `rdf:subject`, `rdf:predicate` and `rdf:object` properties because every RDF parser knows what they are. Push comes to shove, we could actually suggest more friendly labels in the profile while keeping the properties fully functional in an RDF context.

```
{
   "@id": "http://www.credentialengineregistry.org/resources/ce-[UUID]",
   "@type": "ceterms:ThirdPartyAssertion",
   "ceterms:name": "My advanced standing assertion",
   "ceterms:description": "The [credential 1] credential grants advanced standing for [credential 2].",
   "ceterms:assertedBy": { "@id": "http://www.credentialengineregistry.org/resources/ce-[UUID]" },
   "rdf:subject": { "@id": "http://www.credentialengineregistry.org/resources/ce-[UUID]" },
   "rdf:predicate": { "@id": "http://purl.org/ctdl/terms/isAdvancedStandingFor" },
   "rdf:object": { "@id": "http://www.credentialengineregistry.org/resources/ce-[UUID]" }
}
```

**Comment by stuartasutton** [comment]
@siuc-nate, I've carried this one step further by using the rdf:Statement class to split out the triple. Now, a `ceterms:ThirdPartyAssertion` can contain an array of related mappings (triples):

### Turtle
```
@prefix ceterms: <http://purl.org/ctdl/terms/> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix xml: <http://www.w3.org/XML/1998/namespace> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .

<http://www.credentialengineregistry.org/resources/ce-8843b8d0-0236-4e8b-b8e2-7a81fbddcb0e> a ceterms:ThirdPartyAssertion ;
    ceterms:name "My advanced standing assertion" ;
    ceterms:assertedBy <http://www.credentialengineregistry.org/resources/ce-aab560d4-d1b2-4c71-a11a-6f1b9bc65875> ;
    ceterms:description "The [credential 1] credential grants advanced standing for [credential 2]." ;
    ceterms:map [ a rdf:Statement ;
            rdf:object <http://www.credentialengineregistry.org/maps/ce-d3976f68-1833-4524-b602-99baaa46cc21> ;
            rdf:predicate ceterms:isAdvancedStandingFor ;
            rdf:subject <http://www.credentialengineregistry.org/maps/ce-334b4732-058e-43da-a177-9bb9986e0e75> ] .
```
### JSONLD
```
{
  "@context": {
    "ceterms": "http://purl.org/ctdl/terms/",
    "rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#",
    "rdfs": "http://www.w3.org/2000/01/rdf-schema#",
    "xsd": "http://www.w3.org/2001/XMLSchema#"
  },
  "@graph": [
    {
      "@id": "http://www.credentialengineregistry.org/resources/ce-8843b8d0-0236-4e8b-b8e2-7a81fbddcb0e",
      "@type": "ceterms:ThirdPartyAssertion",
      "ceterms:assertedBy": {
        "@id": "http://www.credentialengineregistry.org/resources/ce-aab560d4-d1b2-4c71-a11a-6f1b9bc65875"
      },
      "ceterms:description": "The [credential 1] credential grants advanced standing for [credential 2].",
      "ceterms:map": {
        "@id": "_:ub14bL10C17"
      },
      "ceterms:name": "My advanced standing assertion"
    },
    {
      "@id": "_:ub14bL10C17",
      "@type": "rdf:Statement",
      "rdf:object": {
        "@id": "http://www.credentialengineregistry.org/maps/ce-d3976f68-1833-4524-b602-99baaa46cc21"
      },
      "rdf:predicate": {
        "@id": "ceterms:isAdvancedStandingFor"
      },
      "rdf:subject": {
        "@id": "http://www.credentialengineregistry.org/maps/ce-334b4732-058e-43da-a177-9bb9986e0e75"
      }
    }
  ]
}
```

**Comment by stuartasutton** [comment]
This sort of 3rd-party statements declaring a relationship between two resources --neither of which 'belong' to the 3rd-party-- is actual an outcome of a separate sort of mapping service. Both CASS and ASN have such mapping services. Here's an image from an ASN presentation:

<img width="1308" alt="screenshot 2017-05-30 05 30 49" src="https://cloud.githubusercontent.com/assets/2939046/26583249/5235d1ca-44f9-11e7-8272-a45658dab529.png">

**Comment by siuc-nate** [comment]
Setting aside for a moment where such a service may reside, have we figured out a solution for this in terms of CTDL?

**Comment by siuc-nate** [comment]
@stuartasutton you created some additional graphs to show a proposed solution for this issue. Can you post them here?

Also, we need to discuss this on Tuesday.

**Comment by siuc-nate** [comment]
This issue has been split into:
- Third party mapping: #413 
- Advanced Standing considerations related to the Indiana/Military use case: #432 

I'm closing this for now.
